[SPARK-2883] [SQL] ORC data source for Spark SQL

liancheng opened the pull request:
This PR updates PR #6135 authored by @zhzhan from Hortonworks.

---

This PR implements a Spark SQL data source for accessing ORC files.

> **NOTE**
> 
> Although ORC is now an Apache TLP, the codebase is still tightly coupled with Hive.  That's why the new ORC data source is under `org.apache.spark.sql.hive` package, and must be used with `HiveContext`.  However, it doesn't require existing Hive installation to access ORC files.
## New Features
1.  Saving/loading ORC files without contacting Hive metastore
2.  Support for complex data types (i.e. array, map, and struct)
3.  Aware of common optimizations provided by Spark SQL:
   - Column pruning
   - Partitioning pruning
   - Filter push-down
## Future Work
1.  Schema evolution support
2.  Hive metastore table conversion
## Acknowledgements

This PR also include initial work done by @scwf from Huawei (PR #3753).

zhzhan committed:
orc data source support

zhzhan committed:
minor change

zhzhan committed:
predicate fix

zhzhan committed:
reuse test suite

zhzhan committed:
save mode fix

zhzhan committed:
resolve review comments

zhzhan committed:
resolve review comments

liancheng committed:
Polishes the ORC data source

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32839 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32839/consoleFull) for   PR 6194 at commit [`4bc937f`](https://github.com/apache/spark/commit/4bc937fa37c2674c007726f6c9bb25911378049f).

liancheng commented:
@zhzhan Here is a rough list of my updates:

1.  Rebased to PR #6150, which updated the newly introduced partitioning support API

    Made corresponding changes according to the new API changes.

1.  `OrcFilters` update

    - I worked around the builder state inconsistency issue by employing a double-checking mechanism.  Now we can convert a single child of an `And` filter even if the other child is inconvertible.
    - Added data type checking, as ORC doesn't accept all Spark SQL atomic data types (e.g. timestamp).

1.  Added new tests

    Extracted some useful testing utility methods to `SQLTestUtils`, and added `NewOrcQuerySuite` based on `ParquetQuerySuite`.

1.  Some regular refactoring

    Mainly styling issues and code simplifications.

liancheng commented:
Some TODO items related to testing:

- [ ] Cleanup current ORC test suites, as most of them are based on old Parquet test code, which has been deprecated and removed.
- [ ] More tests on filter push-down

zhzhan commented:
@liancheng  Thanks for the followup. For the future work, feel free to assign to me.

zhzhan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala:
```diff
@@ -0,0 +1,144 @@
+          .orElse(tryLeft.flatMap(_ => buildSearchArgument(left, builder)))
+          .orElse(tryRight.flatMap(_ => buildSearchArgument(right, builder)))
+
+      case And(left, right) =>
```
Should be Or?

liancheng replied:
Oops, thanks!

SparkQA commented:
[Test build #32839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32839/consoleFull) for   PR 6194 at commit [`4bc937f`](https://github.com/apache/spark/commit/4bc937fa37c2674c007726f6c9bb25911378049f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32839/
Test FAILed.

zhzhan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFilters.scala:
```diff
@@ -0,0 +1,144 @@
+    // children with brand new builders, and only do the actual conversion with the right builder
+    // instance when the children are proven to be convertible.
+    //
+    // P.S.: Hive seems to use `SearchArgument` together with `ExprNodeGenericFuncDesc` only.
```
I checked with hive team. For external user, it is more expected to use the current builder  approach, although hive internally build xml file by ExprNodeGenericFuncDesc.

liancheng replied:
Thanks. Do you know are there any other projects that uses ORC `SearchArgument` builder API? I'm looking for examples. I think the problem we faced should be pretty general. Would like to see how other projects solve it.

scwf replied:
is SearchArgument builder API stable/compatible for different hive version?

liancheng replied:
@scwf Good question. @zhzhan Would you mind helping confirming maturity of this API?

liancheng replied:
@scwf BTW, with the help of the newly introduced isolated classloader mechanism, Spark SQL can always depend on the most recent version of Hive. At the meanwhile, users can specify arbitrary Hive metastore version to use. So even if this API changes across Hive versions, we don't need shim code to ensure compatibility.

scwf replied:
get it, thanks for the explanation

zhzhan commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32848 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32848/consoleFull) for   PR 6194 at commit [`4bc937f`](https://github.com/apache/spark/commit/4bc937fa37c2674c007726f6c9bb25911378049f).

zhzhan commented:
@liancheng FYI: For schema merging, I checked some orc experts, and probably it is not supported the filter push down if the column is not in that specific orc file (I myself does not check the implementation yet). In the meantime, separating orc from hive is a on-going effort. We can separate orc from hive afterwards, and upgrade orc support to latest, which I think will improve the performance and a lot and remove potential version mismatch due to hive versions.

SparkQA commented:
[Test build #32848 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32848/consoleFull) for   PR 6194 at commit [`4bc937f`](https://github.com/apache/spark/commit/4bc937fa37c2674c007726f6c9bb25911378049f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32848/
Test PASSed.

liancheng committed:
ORC filter bug fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32874/consoleFull) for   PR 6194 at commit [`563ee1a`](https://github.com/apache/spark/commit/563ee1a8641156bdc529a26a060fdc877bc84d3f).

liancheng commented:
@zhzhan Thanks for the information.

liancheng committed:
Adds @since and @Experimental annotations

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32881 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32881/consoleFull) for   PR 6194 at commit [`eda453d`](https://github.com/apache/spark/commit/eda453db79beeb446f21b9e536e00a560984e888).

SparkQA commented:
[Test build #32874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32874/consoleFull) for   PR 6194 at commit [`563ee1a`](https://github.com/apache/spark/commit/563ee1a8641156bdc529a26a060fdc877bc84d3f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  implicit class OrcContext(sqlContext: HiveContext) `
  * `  implicit class OrcDataFrame(dataFrame: DataFrame) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32874/
Test PASSed.

scwf commented on sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala:
```diff
@@ -143,6 +145,9 @@ private[sql] class SQLConf extends Serializable with CatalystConf {
     getConf(PARQUET_USE_DATA_SOURCE_API, "true").toBoolean
 
+  private[spark] def orcFilterPushDown =
+    getConf(ORC_FILTER_PUSHDOWN_ENABLED, "false").toBoolean
```
why this is false by default?

liancheng replied:
Filter push-down is a pretty error prone feature, 'cause it requires different components to interoperate seamlessly. We'd rather disable it by default until it's proven stable.

rxin replied:
How flaky is this? I'd maybe turn it on by default during QA, and then consider disabling it when we release if it is indeed flaky.

scwf commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/HadoopTypeConverter.scala:
```diff
@@ -0,0 +1,61 @@
+  /**
+   * Wraps with Hive types based on object inspector.
+   */
+  def wrappers(oi: ObjectInspector): Any => Any = wrapperFor(oi)
```
how about add this two method to HiveInspectors instead of making a new object.

zhzhan replied:
@scwf good point. Actually after moving the first method to HiveInspectors, TableReader can use it as well. But doe it make sense to have a separate PR to do it, as this PR scope is quite limited to orc package, and has zero impact on any other features.

liancheng replied:
Thanks, moved them.

rxin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala:
```diff
@@ -0,0 +1,74 @@
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.{DataFrame, SaveMode}
+
+package object orc {
```
why having this? this is not much more concise than 

sqlContext.load("orc", paths...)

zhzhan replied:
@rxin The code has been written for a while, before the new external data source API. Parquet and ORC are both very popular format. With this two API, user  operate on orc format quite similar to parquet format, which may be more convenient to end user.

rxin replied:
What I'm saying is that in terms of total characters users have to type, it is actually strictly more to have this implicit conversion. Also implicit conversions tend to bring problems, impacting debuggability, binary compatibility, etc. We should just remove this.

liancheng replied:
@rxin I removed them and added "orc" as a built-in data source. Also migrated other parts to the new DataFrame reader/writer API.

SparkQA commented:
[Test build #32881 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32881/consoleFull) for   PR 6194 at commit [`eda453d`](https://github.com/apache/spark/commit/eda453db79beeb446f21b9e536e00a560984e888).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  implicit class OrcContext(sqlContext: HiveContext) `
  * `  implicit class OrcDataFrame(dataFrame: DataFrame) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32881/
Test PASSed.

tedyu commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/package.scala:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * ::Experimental::
+     *
+     * Loads specified Parquet files, returning the result as a [[DataFrame]].
```
Parquet -> ORC

liancheng replied:
Oops... Thanks for pointing out the typo, but I just removed this file in favor of the new DataFrame reader/writer API :)

liancheng committed:
Addresses comments

- Migrates to the new DataFrame reader/writer API
- Merges HadoopTypeConverter into HiveInspectors
- Refactors test suites

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32907/consoleFull) for   PR 6194 at commit [`d4afeed`](https://github.com/apache/spark/commit/d4afeed86f2094df8cb6ba509f5a2da22c3bf02b).

SparkQA commented:
[Test build #32907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32907/consoleFull) for   PR 6194 at commit [`d4afeed`](https://github.com/apache/spark/commit/d4afeed86f2094df8cb6ba509f5a2da22c3bf02b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `              sys.error(s"Failed to load class for data source: $provider")`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32907/
Test PASSed.

liancheng committed:
Reorganizes ORC test suites

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32926 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32926/consoleFull) for   PR 6194 at commit [`55ecd96`](https://github.com/apache/spark/commit/55ecd9641838ce4af364005917dcd1d0ffadc3b6).

rxin commented:
LGTM with respect to API change (there isn't any).

SparkQA commented:
[Test build #32926 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32926/consoleFull) for   PR 6194 at commit [`55ecd96`](https://github.com/apache/spark/commit/55ecd9641838ce4af364005917dcd1d0ffadc3b6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `              sys.error(s"Failed to load class for data source: $provider")`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32926/
Test PASSed.

liancheng commented:
In the last a few commits, I added `"orc"` as a built-in data source name, so that we can have

```scala
hiveContext.read.format("orc").load("hdfs://...")
```

and

```scala
df.write.format("orc").save("hdfs://...")
```

Note that ORC data source is coupled with Hive. If users try to use it with `SQLContext`, an error message will be thrown to ask users to use `HiveContext` instead.

liancheng commented:
@marmbrus This should be ready to go.

marmbrus commented on sql/core/src/main/scala/org/apache/spark/sql/sources/ddl.scala:
```diff
@@ -188,18 +188,20 @@ private[sql] class DDLParser(
+    "jdbc" -> "org.apache.spark.sql.jdbc.DefaultSource",
+    "json" -> "org.apache.spark.sql.json.DefaultSource",
+    "parquet" -> "org.apache.spark.sql.parquet.DefaultSource",
+    "orc" -> "org.apache.spark.sql.hive.orc.DefaultSource"
```
Should we give a nice error here if not compiled with hive support?

marmbrus commented:
Thanks guys!  Merged to master and 1.4.
